### PR TITLE
nixos/jupyter: add env kernel option

### PR DESCRIPTION
**Pull request by jluttine**
###### Motivation for this change

Jupyter kernels support `env` option in order to define environment variables for kernels: https://jupyter-client.readthedocs.io/en/stable/kernels.html

This PR adds support for it in NixOS, so that, for instance, in JupyterHub one can define multiple kernels with different environment variables. Below is a minimal example:

```
jupyterhub = {
  enable = true;
  kernels = {
    python3 = let
      env = pkgs.python3.withPackages (ps: with ps; [
        ipykernel
        numpy
      ]);
    in {
      displayName = "Python with NumPy";
      argv = [
        "${env.interpreter}"
        "-m"
        "ipykernel_launcher"
        "-f"
        "{connection_file}"
      ];
      # This PR adds support for this option:
      env = {
        OMP_NUM_THREADS = "1";
      };
      language = "python";
      logo32 = "${env}/${env.sitePackages}/ipykernel/resources/logo-32x32.png";
      logo64 = "${env}/${env.sitePackages}/ipykernel/resources/logo-64x64.png";
    };
  };
};
```

I tested this on my machine and it worked.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jluttine** [comment]
I rebased the branch on nixos-unstable. This PR is still relevant.

**Comment by ncfavier** [comment]
There are other possible kernel settings that don't have matching options:

https://github.com/NixOS/nixpkgs/blob/ba187fbdc5e35322c7dff556ef2c47bddfd6e8d7/pkgs/applications/editors/jupyter/kernel.nix#L42

We should probably add them all or give the submodule a JSON `freeformType` following [RFC 42](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md#part-1-1) (possibly both).

File: nixos/modules/services/development/jupyter/kernel-options.nix [hunk 1]
```diff
@@ -40,6 +40,17 @@ with lib;
       '';
     };
 
+    env = mkOption {
+      type = types.attrsOf types.str;
+      default = {};
+      example = {
+        OMP_NUM_THREADS = "1";
+      };
+      description = ''
```
Comment (ncfavier) [inline nixos/modules/services/development/jupyter/kernel-options.nix:49]:
```suggestion
      description = lib.mdDoc ''
```

**Comment by jluttine** [comment]
I tried to define `freeformType`, but I'm now getting the following error:

```
Cacheable portion of option doc build failed.
Usually this means that an option attribute that ends up in documentation (eg `default` or `description`) depends on the restricted module arguments `config` or `pkgs`.

Rebuild your configuration with `--show-trace` to find the offending location. Remove the references to restricted arguments (eg by escaping their antiquotations or adding a `defaultText`) or disable the sandboxed build for the failing module by setting `meta.buildDocsInSandbox = false`.

error: builder for '/nix/store/l8j6jn4dv2mcnb0dr9zzwxm7c8f4qslx-lazy-options.json.drv' failed with exit code 1
```

I could make it work by defining `freeformType = ...` outside kernel-options.nix and just merging it to the resulting attribute set after importing the file. But I would need to do that in two places because the file is imported in two places. Also, it feels wrong to define it outside kernel-options.nix because it should belong there. Any ideas what I'm doing wrong and how to fix this?

I don't have a strong opinion whether `env` should be defined as a NixOS option or just drop that commit and let it be supported with this `freeformType` solution. I'm ok with either way.

File: nixos/modules/services/development/jupyter/kernel-options.nix [hunk 2]
```diff
@@ -1,11 +1,13 @@
...
-{lib }:
+{ lib, pkgs }:
 
 with lib;
 
 {
   options = {
 
+    freeformType = (pkgs.formats.json { }).type;
+
```
Comment (ncfavier) [inline nixos/modules/services/development/jupyter/kernel-options.nix:10]:
```suggestion
  freeformType = (pkgs.formats.json { }).type;

  options = {

```

**Comment by ncfavier** [comment]
> I don't have a strong opinion whether `env` should be defined as a NixOS option or just drop that commit and let it be supported with this `freeformType` solution. I'm ok with either way.

I think it's fine to keep it.

File: nixos/modules/services/development/jupyter/kernel-options.nix (hunk 2 above)
Comment (jluttine) [reply nixos/modules/services/development/jupyter/kernel-options.nix]: Thanks, that worked! :+1:

**Review by ncfavier — APPROVED** [review]

**Review by RaitoBezarius — APPROVED** [review]

**Pull request merged**